delta_Go/game_mechanics.py

```diff
@@ -0,0 +1,260 @@
+            )
+
+        return self.observation, 0, self.done, self.info
+
+    def move_to_string(self, move: int):
+        N = self.observation.shape[0]
+        if move == N**2:
+            return "passes"
+        return f"places counter at coordinate: {(move//N, move%N)}"
+
+    def __str__(self):
+        return str(self.observation) + "\n"
+
+    def _step(self, move: int) -> float:
+
+        assert not self.done, "Game is done! Please reset() the env before calling step() again"
+        assert move in self.legal_moves, f"{move} is an illegal move"
+
+        if self.verbose:
+            print(f"{self.turn_pretty} {self.move_to_string(move)}")
+
+        self.env.step(move)
+
+        if self.render:
+            self.render_game()
+
+        return self.reward
+
+    @property
+    def reward(self):
+        return self.env.last()[1]
+
+    @property
+    def player_score(self):
+        black_score = self.env.env.env.env.go_game.score()  # lol
+        return black_score if self.player == "black_0" else black_score * -1
+
+    def step(self, move: int):
+
+        # Flipped because the env takes the step, changes the player, then we return the reward
+        reward = -self._step(move)
+
+        if not self.done:
+            opponent_reward = self._step(
+                self.opponent_choose_move(
+                    observation=self.observation, legal_moves=self.legal_moves, env=self
+                ),
+            )
+            # Flipped as above
+            reward = opponent_reward
+
+        if self.verbose and self.done:
+            white_idx = int(self.turn_pretty == "white")
+            black_idx = int(self.turn_pretty == "black")
+            white_count = np.sum(self.env.last()[0]["observation"].astype("int")[:, :, white_idx])
+            black_count = np.sum(self.env.last()[0]["observation"].astype("int")[:, :, black_idx])
+
+            print(
+                f"\nGame over. Reward = {reward}.\n"
+                f"Player was playing as {self.player[:-2]}.\n"
+                f"White has {white_count} counters.\n"
+                f"Black has {black_count} counters.\n"
+                f"Your score is {self.player_score}.\n"
+            )
+
+        return self.observation, reward, self.done, self.info
+
+
+def GoEnv(
+    opponent_choose_move: Callable[[np.ndarray, np.ndarray], int],
+    verbose: bool = False,
+    render: bool = False,
+    game_speed_multiplier: int = 1,
+) -> DeltaEnv:
+    return DeltaEnv(
+        go_v5.env(board_size=BOARD_SIZE, komi=KOMI),
+        opponent_choose_move,
+        verbose,
+        render,
+        game_speed_multiplier=game_speed_multiplier,
+    )
+
+
+def choose_move_randomly(observation, legal_moves, env=None):
+    return random.choice(legal_moves)
+
+
+def choose_move_pass(observation, legal_moves, env) -> int:
+    """passes on every turn."""
+    return BOARD_SIZE**2
+
+
+def transition_function(env: DeltaEnv, action: int) -> DeltaEnv:
+    env = deepcopy(env)
```

Comment:
Add a TODO with a jira ticket number in it for speeding this up?

Might not be necessary to actually do this in a while (since we might run chess next time we do Intermediate?)